### PR TITLE
Software FEC attempt 2

### PR DESCRIPTION
**Pull request by mcm001**
See also:
https://www.ti.com/lit/an/swra313/swra313.pdf
https://www.ti.com/lit/an/swra113a/swra113a.pdf

### CODE REVIEW

**Comment by mcm001** [comment]
At 48mhz on my stm32 here I get these times for a 48 byte long packet + 2 byte crc. Not super fast, but it'll hit 50hz telemetry at least I think. Hanging for 9ms on every radio message is not ideal though lol. And that's with -Ofast
image

**Comment by mcm001** [comment]
image
and at 84mhz. decode (RX) time is linear with cpu speed

File: common/utils/ti_fec.h [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+#include <stddef.h>
+#include <stdint.h>
+
+template <size_t MessageLen>
+class FecEncoder {
+ public:
+  FecEncoder() = default;
+
+  // Encode a message. CRC is assumed to have already been added.
+  uint8_t* Encode(uint8_t* input);
```
Comment (mcm001) [inline common/utils/ti_fec.h:12]: The fact that this returns a pointer to a member means you gotta be extra careful that the class instance sticks around. I don't love this.

File: ci/autoformat.sh [hunk 2]
```diff
@@ -11,4 +11,4 @@ do
 done
 
 # Cpplint
-cpplint --counting=detailed --recursive common/ desktop_projects/ stm32_projects/device_drivers/ platformio_projects/src/ platformio_projects/lib/device_drivers/ platformio_projects/lib/new_hal/
+cpplint --counting=detailed --filter="-build/c++11" --recursive common/ desktop_projects/ stm32_projects/device_drivers/ platformio_projects/src/ platformio_projects/lib/device_drivers/ platformio_projects/lib/new_hal/
```
Comment (SamHaggans) [inline ci/autoformat.sh:14]: What issue does this fix? Assuming some warning we don't want?

File: common/utils/ti_fec.cpp [hunk 3]
```diff
@@ -0,0 +1,28 @@
+#include "ti_fec.h"
```
Comment (SamHaggans) [inline common/utils/ti_fec.cpp:1]: Can you rearrange/rename some of the `ti_fec*` files? For the ones that are classes, it should probably be one header file for the class and then a source file. It looks like there's also some general utilities, which can be its own header/source. The current naming scheme doesn't make it immediately clear what is located where.

File: common/utils/ti_fec.h [hunk 4]
```diff
@@ -0,0 +1,108 @@
+#pragma once
+
+#include <stddef.h>
+#include <stdint.h>
+
+#include "board_config_common.h"
+
+#define DEBUG_PRINTF(a...) \
```
Comment (SamHaggans) [inline common/utils/ti_fec.h:8]: Probably should be removed, if just for debugging.

File: common/utils/ti_fec.h [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+   * @param pInputMessage where to find the FEC encoded message
+   * @param pOutputBuffer where to store the decoded message
+   * @param decodedMessageLen How long we expect the decoded message to be.
+   */
+  void FecDecode(uint8_t* pInputMessage, uint8_t* pOutputBuffer,
+                 size_t decodedMessageLen);
+
+ private:
+  /**************************************************************************************************************
+   * @fn fecDecode
```
Comment (SamHaggans) [inline common/utils/ti_fec.h:67]: Your docstrings don't use a consistent format. We don't really have a good standard elsewhere but probably worth keeping it consistent in the same file.

File: common/utils/ti_fec_decode.cpp [hunk 6]
```diff
@@ -0,0 +1,318 @@
+#include <stdint.h>
+
+#include <algorithm>
+#include <array>
+#include <cstdio>
+#include <cstring>
+#include <queue>
+#include <random>
+
+// #include <chrono>
```
Comment (SamHaggans) [inline common/utils/ti_fec_decode.cpp:10]: Remove commented debugging things. A few other in this file, too.

File: desktop_projects/pybind_test/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,17 @@
+add_library(stm32-avionics-py SHARED
```
Comment (SamHaggans) [inline desktop_projects/pybind_test/CMakeLists.txt:1]: Adding pybind is probably out of scope for this PR. Not that it's bad, but there's already a lot here so it'd probably be worth separating it for a future change.

File: common/system/hardware_manager.c [hunk 8]
```diff
@@ -958,7 +959,7 @@ void hm_pyroSetPwm(int pyroId, uint32_t duration, uint32_t frequency,
 #endif  // HAS_DEV(PYRO_DIGITAL)
 }
 
-void hm_pyroUpdate() {
+void hm_pyroUpdate(void *pUserData) {
```
Comment (SamHaggans) [inline common/system/hardware_manager.c:962]: Out of scope here, but probably worth refactoring some of our raw callbacks if it means that we're having to add unused parameters to avoid warnings.

File: stm32_projects/amazon-stm32-arduino/Core/Src/main.cpp [hunk 9]
```diff
@@ -0,0 +1,300 @@
...
+/* USER CODE BEGIN PFP */
+
+#include <stdint.h>
+
+/* USER CODE END PFP */
+
+/* Private user code ---------------------------------------------------------*/
+/* USER CODE BEGIN 0 */
+
+static uint16_t calculateCRC(uint8_t crcData, uint16_t crcReg) {
```
Comment (SamHaggans) [inline stm32_projects/amazon-stm32-arduino/Core/Src/main.cpp:64]: We've used the `amazon-stm32-arduino` project for introductory things in the past. Do we want to make a separate cube project for the crc testing so that we can keep the original one? Maybe easier to just make a new one, though, and can be in a separate PR.

File: stm32_projects/amazon-stm32-arduino/Release/Core/Src/gpio.cyclo [hunk 10]
```diff
@@ -0,0 +1 @@
+gpio.c:41:6:MX_GPIO_Init	1
```
Comment (SamHaggans) [inline stm32_projects/amazon-stm32-arduino/Release/Core/Src/gpio.cyclo:1]: What do these `.cyclo` files do?

File: stm32_projects/fcb_v0/Core/Src/main.cpp [hunk 11]
```diff
@@ -61,6 +61,10 @@ void SystemClock_Config(void);
 void PeriphCommonClock_Config(void);
 /* USER CODE BEGIN PFP */
 
+//#include <type_traits>
```
Comment (SamHaggans) [inline stm32_projects/fcb_v0/Core/Src/main.cpp:64]: I think this can be removed.

File: stm32_projects/device_drivers/ti_radio.cpp [hunk 12]
```diff
@@ -16,6 +16,18 @@
...
+#include "ti_fec.h"
+
+#define TI_RADIO_STATUS_MASK 0xF0
+
+#pragma GCC optimize("-O2")
+
+// Static fec encode/decode helpers
+// This assumes no multi-threading weirdness
+#if RADIO_TI_TYPE == RADIO_TI_TYPE_CC1120
+FecEncoder encoder;
```
Comment (SamHaggans) [inline stm32_projects/device_drivers/ti_radio.cpp:28]: Should these be `static`?

File: ci/autoformat.sh (hunk 2 above)
Comment (mcm001) [reply ci/autoformat.sh:14]: cpplint is stuck on C++11, so we have to suppress or just dont include crono, ever

File: stm32_projects/amazon-stm32-arduino/Core/Src/main.cpp (hunk 9 above)
Comment (mcm001) [reply stm32_projects/amazon-stm32-arduino/Core/Src/main.cpp:64]: I can copy this code into a FEC benchmark project, and revert changes to amazon-stm-arduino

File: common/utils/ti_fec.h (hunk 4 above)
Comment (mcm001) [reply common/utils/ti_fec.h:8]: It's also for profiling? IDK, i think it's worth keeping as a empty macro

File: common/system/hardware_manager.c (hunk 8 above)
Comment (mcm001) [reply common/system/hardware_manager.c:962]: Tracked in #268
